File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/saveable/CacheableListModelSaver.java
package com.raizlabs.android.dbflow.sql.saveable;

import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.structure.ModelAdapter;
import com.raizlabs.android.dbflow.structure.database.DatabaseStatement;
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper;

import java.util.Collection;

/**
 * Description: Used for model caching, enables caching models when saving in list.
 */
public class CacheableListModelSaver<TModel>
    extends ListModelSaver<TModel> {

    public CacheableListModelSaver(@NonNull ModelSaver<TModel> modelSaver) {
        super(modelSaver);
    }

    @Override
    public synchronized void saveAll(@NonNull Collection<TModel> tableCollection,
                                     @NonNull DatabaseWrapper wrapper) {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return;
        }

        ModelSaver<TModel> modelSaver = getModelSaver();
        ModelAdapter<TModel> modelAdapter = modelSaver.getModelAdapter();
        DatabaseStatement statement = modelAdapter.getInsertStatement(wrapper);
        DatabaseStatement updateStatement = modelAdapter.getUpdateStatement(wrapper);
        try {
            for (TModel model : tableCollection) {
                if (modelSaver.save(model, wrapper, statement, updateStatement)) {
                    modelAdapter.storeModelInCache(model);
                }
            }
        } finally {
            updateStatement.close();
            statement.close();
        }
    }

    @Override
    public synchronized void insertAll(@NonNull Collection<TModel> tableCollection,
                                       @NonNull DatabaseWrapper wrapper) {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return;
        }

        ModelSaver<TModel> modelSaver = getModelSaver();
        ModelAdapter<TModel> modelAdapter = modelSaver.getModelAdapter();
        DatabaseStatement statement = modelAdapter.getInsertStatement(wrapper);
        try {
            for (TModel model : tableCollection) {
                if (modelSaver.insert(model, statement, wrapper) > 0) {
                    modelAdapter.storeModelInCache(model);
                }
            }
        } finally {
            statement.close();
        }
    }

    @Override
    public synchronized void updateAll(@NonNull Collection<TModel> tableCollection,
                                       @NonNull DatabaseWrapper wrapper) {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return;
        }
        ModelSaver<TModel> modelSaver = getModelSaver();
        ModelAdapter<TModel> modelAdapter = modelSaver.getModelAdapter();
        DatabaseStatement statement = modelAdapter.getUpdateStatement(wrapper);
        try {
            for (TModel model : tableCollection) {
                if (modelSaver.update(model, wrapper, statement)) {
                    modelAdapter.storeModelInCache(model);
                }
            }
        } finally {
            statement.close();
        }
    }

    @Override
    public synchronized void deleteAll(@NonNull Collection<TModel> tableCollection,
                                       @NonNull DatabaseWrapper wrapper) {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return;
        }

        ModelSaver<TModel> modelSaver = getModelSaver();
        for (TModel model : tableCollection) {
            if (modelSaver.delete(model, wrapper)) {
                getModelSaver().getModelAdapter().removeModelFromCache(model);
            }
        }
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/saveable/CacheableListModelSaver.kt
package com.raizlabs.android.dbflow.sql.saveable

import com.raizlabs.android.dbflow.structure.database.DatabaseStatement
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper

/**
 * Description: Used for model caching, enables caching models when saving in list.
 */
class CacheableListModelSaver<T : Any>(modelSaver: ModelSaver<T>) : ListModelSaver<T>(modelSaver) {

    @Synchronized override fun saveAll(tableCollection: Collection<T>,
                                       wrapper: DatabaseWrapper): Long {
        val modelAdapter = modelSaver.modelAdapter
        val statement = modelAdapter.getInsertStatement(wrapper)
        val updateStatement = modelAdapter.getUpdateStatement(wrapper)
        return applyAndCount(tableCollection, statement, updateStatement,
                modelSaver.modelAdapter::storeModelInCache) {
            modelSaver.save(it, statement, updateStatement, wrapper)
        }
    }

    @Synchronized override fun insertAll(tableCollection: Collection<T>,
                                         wrapper: DatabaseWrapper): Long {
        val statement = modelSaver.modelAdapter.getInsertStatement(wrapper)
        return applyAndCount(tableCollection, statement,
                cacheFn = modelSaver.modelAdapter::storeModelInCache) {
            modelSaver.insert(it, statement, wrapper) > 0
        }
    }

    @Synchronized override fun updateAll(tableCollection: Collection<T>,
                                         wrapper: DatabaseWrapper): Long {
        val statement = modelSaver.modelAdapter.getUpdateStatement(wrapper)
        return applyAndCount(tableCollection, statement,
                cacheFn = modelSaver.modelAdapter::storeModelInCache) {
            modelSaver.update(it, statement, wrapper)
        }
    }

    @Synchronized override fun deleteAll(tableCollection: Collection<T>,
                                         wrapper: DatabaseWrapper): Long {
        val statement = modelSaver.modelAdapter.getDeleteStatement(wrapper)
        return applyAndCount(tableCollection, statement,
                cacheFn = modelSaver.modelAdapter::removeModelFromCache) {
            modelSaver.delete(it, statement, wrapper)
        }
    }

    private inline fun applyAndCount(tableCollection: Collection<T>,
                                     databaseStatement: DatabaseStatement,
                                     otherStatement: DatabaseStatement? = null,
                                     crossinline cacheFn: (T) -> Unit,
                                     crossinline fn: (T) -> Boolean): Long {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return 0L
        }

        var count = 0L
        try {
            tableCollection.forEach {
                if (fn(it)) {
                    cacheFn(it)
                    count++
                }
            }
        } finally {
            databaseStatement.close()
            otherStatement?.close()
        }
        return count
    }
}
